router/core/src/main/scala/com/twitter/finagle/buoyant/RetryFilter.scala

```diff
@@ -0,0 +1,85 @@
+package com.twitter.finagle.buoyant
+
+import com.twitter.conversions.time._
+import com.twitter.finagle.service.{RetryBudget, RetryPolicy}
+import com.twitter.finagle.stats.StatsReceiver
+import com.twitter.finagle.tracing.Trace
+import com.twitter.finagle.{Filter, Service}
+import com.twitter.util._
+
+/**
+ * Cribbed from:
+ * https://github.com/twitter/finagle/blob/develop/finagle-core/src/main/scala/com/twitter/finagle/service/RetryFilter.scala
+ *
+ * Copied here and modified in order to change the type of stats exported by the filter.
+ */
+class RetryFilter[Req, Rep](
+  retryPolicy: RetryPolicy[(Req, Try[Rep])],
+  timer: Timer,
+  statsReceiver: StatsReceiver,
+  retryBudget: RetryBudget
+)
+  extends Filter[Req, Rep, Req, Rep] {
+
+  def this(
+    retryPolicy: RetryPolicy[(Req, Try[Rep])],
+    timer: Timer,
+    statsReceiver: StatsReceiver
+  ) = this(
+    retryPolicy,
+    timer,
+    statsReceiver,
+    RetryBudget()
+  )
+
+  private[this] val retriesStat = statsReceiver.stat("retries")
+
+  private[this] val totalRetries = statsReceiver.scope("retries").counter("total")
+
+  private[this] val budgetExhausted =
+    statsReceiver.scope("retries").counter("budget_exhausted")
```

Comment:
I'm a bit torn on this change. As currently written, our modifications to the stats exported by this filter are additive -- somebody using our filter instead of Finagle's would get all of the original Finagle stats plus one more.  If we rescope the retries stat though, that won't be the case.  Happy to make the change, just wanted to mention why it's written as it is.